Mathlib/LinearAlgebra/RootSystem/OfBilinear.lean

```diff
@@ -0,0 +1,194 @@
+this construction, replacing the ring of integers with an arbitrary commutative ring and the
+integral lattice with an arbitrary reflexive module equipped with a bilinear form.
+
+## Main definitions:
+ * `IsReflective`: Length is a regular value of `R`, and reflection is definable.
+ * `coroot_of_reflective`: The coroot corresponding to a reflective vector.
+ * `of_Bilinear`: The root pairing whose roots are reflective vectors.
+
+## TODO
+ * properties
+-/
+
+open Set Function Module
+
+noncomputable section
+
+variable {ι R M N : Type*} [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+namespace RootPairing
+
+section Construction
+
+/-- A vector `x` is reflective with respect to a bilinear form if multiplication by its norm is
+injective, and for any other vector `y`, there is a scalar that takes the norm of `x` to twice the
+inner product of `x` and `y`. These conditions are what we need when describing reflection as a map
+taking `y` to `y - 2 • (B x y) / (B x x) • x`. -/
+structure IsReflective (B : M →ₗ[R] M →ₗ[R] R) (x : M) : Prop :=
+  regular : IsRegular (B x x)
+  dvd_two_mul : ∀ y, B x x ∣ 2 * B x y
+
+variable (B : M →ₗ[R] M →ₗ[R] R) {x : M}
+
+lemma isReflective_of_dvd_two [IsDomain R] [NeZero (2 : R)] (hx : B x x ∣ 2) :
+    IsReflective B x where
+  regular := isRegular_of_ne_zero <| fun contra ↦ by simp [contra, two_ne_zero (α := R)] at hx
+  dvd_two_mul y := hx.mul_right (B x y)
+
+/-- The coroot attached to a reflective vector. -/
+def coroot_of_reflective (hx : IsReflective B x) :
+    M →ₗ[R] R where
+  toFun y := (hx.2 y).choose
+  map_add' a b := by
+    refine hx.regular.left ?_
+    simp only
+    rw [← (hx.dvd_two_mul (a + b)).choose_spec, mul_add, ← (hx.2 a).choose_spec,
+      ← (hx.2 b).choose_spec, map_add, mul_add]
+  map_smul' r a := by
+    refine hx.1.1 ?_
+    simp only [RingHom.id_apply]
+    rw [← (hx.2 (r • a)).choose_spec, smul_eq_mul, mul_left_comm, ← (hx.2 a).choose_spec, map_smul,
+      two_mul, smul_eq_mul, two_mul, mul_add]
+
+@[simp]
+lemma coroot_of_reflective_apply {y : M} (hx : IsReflective B x) :
+    B x x * coroot_of_reflective B hx y = 2 * B x y := by
+  dsimp [coroot_of_reflective]
+  rw [← (hx.2 y).choose_spec]
+
+@[simp]
+lemma coroot_of_reflective_apply_self (hx : IsReflective B x) :
+    coroot_of_reflective B hx x = 2 := by
+  refine hx.1.1 ?_
+  dsimp only
+  rw [coroot_of_reflective_apply B hx, mul_two, two_mul]
+
+lemma reflection_reflective_vector {y : M}
+    (hx : IsReflective B x) : Module.reflection (coroot_of_reflective_apply_self B hx) y =
+    y - (coroot_of_reflective B hx y) • x :=
+  rfl
+
+lemma bilinear_apply_reflection_reflective_vector (B : M →ₗ[R] M →ₗ[R] R) (hSB : LinearMap.IsSymm B)
+    {x y z : M} (hx : IsReflective B x) :
+    B (Module.reflection (coroot_of_reflective_apply_self B hx) y)
+      (Module.reflection (coroot_of_reflective_apply_self B hx) z) = B y z := by
```

Comment:
I think it would be nice to state this using the language of orthogonal linear transformations. To my surprise we do not seem to have a definition for these (except possibly in the specialised setting of `InnerProductSpace`).

What do you think about adding something like this somewhere in `Mathlib.LinearAlgebra.SesquilinearForm`:
```lean
variable {R M : Type*} [CommRing R] [AddCommGroup M] [Module R M]
  (B : LinearMap.BilinForm R M) (f : Module.End R M)

def LinearMap.IsOrthogonal : Prop :=
  ∀ x y, B (f x) (f y) = B x y

variable {B f}

@[simp]
lemma LinearEquiv.isAdjointPair_symm_iff {f : M ≃ₗ[R] M} :
    LinearMap.IsAdjointPair B B f f.symm ↔ B.IsOrthogonal f :=
  ⟨fun hf x y ↦ by simpa using hf x (f y), fun hf x y ↦ by simpa using hf x (f.symm y)⟩

lemma LinearMap.isOrthogonal_of_forall_apply_same
    (h : IsLeftRegular (2 : R)) (hB : B.IsSymm) (hf : ∀ x, B (f x) (f x) = B x x) :
    B.IsOrthogonal f := by
  intro x y
  suffices 2 * B (f x) (f y) = 2 * B x y from h this
  -- Would this be better as a `calc` proof? (not sure)
  have := hf (x + y)
  simp only [map_add, add_apply, hf x, hf y, hB.eq (f y), hB.eq y] at this
  simp only [add_assoc, add_right_inj] at this
  simp only [← add_assoc, add_left_inj] at this
  simpa only [← two_mul] using this
```
and then using it to state this lemma as:
```suggestion
lemma isOrthogonal_reflection (B : M →ₗ[R] M →ₗ[R] R) (hSB : B.IsSymm)
    {x : M} (hx : IsReflective B x) :
    B.IsOrthogonal (Module.reflection (corootOfIsReflective_apply_self B hx)) := by
```